Suggested starting point for `deparse`

NikNakk opened the pull request:
There's now the beginnings of a workable `deparse` package with support for factors, lists, data.frames, tibbles. It includes an `as_tribble` option for data.frames and tibbles.

Let me know what you think.

Commit:
Further work on deparse

Created method for list and got the tribble
export working

Commit:
Added further tests and travis

Commit:
Changed expect_equal to expect_identical

Commit:
Added more tribble tests

Also corrected a typo in `deparse.date`

Commit:
Fixed UTF8 character in DESCRIPTION

Also added tibble, dplyr to Suggests

Commit:
Added more tests

Also excluded compat-purrr from codecov

Commit:
Added more tests for srcref_call and data.frame

Also fixed issue where `row.names` where
changed to `character`

Commit:
Added wrapping of list in structure

Also fixed bug in deparse.data.frame and added more tests

Commit:
Added test for `list` with `class`

Commit:
Better handling of lists and tribble

Commit:
Ordered factors now handled properly

Commit:
Tests for `generate_mutate = FALSE`

Also added warning where because factors don't
work well without the `mutate`.

Commit:
Addded test for `ordered`

krlmlr commented on DESCRIPTION:
```diff
@@ -1,7 +1,7 @@
 Title: What the Package Does (one line, title case)
 Version: 0.0-0
-Authors@R: person("Kirill", "Müller", role = c("aut", "cre"), email = "krlmlr+r@mailbox.org")
+Authors@R: person("Kirill", "Muller", role = c("aut", "cre"), email = "krlmlr+r@mailbox.org")
```
We have `Encoding: UTF-8` further below, why do you think the ü is a problem?

NikNakk replied:
It threw invalid multibyte character errors on both travis and appveyor. Not sure how to fix that, so it seemed easiest to replace the ü with a u for now.

krlmlr commented on DESCRIPTION:
```diff
@@ -12,8 +12,13 @@ Date: 2016-08-16
 URLNote: https://github.com/krlmlr/
 BugReports: https://github.com/krlmlr/
 URL: http://krlmlr.github.io/
+Depends:
```
I'd prefer `Imports` here. What do we need the rlang dependency for?

NikNakk replied:
Sorry, should have been `Imports`. It's actually because I've used some of `compat-purrr.R` which has an `rlang` dependency. I could of course use `vapply` etc. or have a `purrr` dependency, but I prefer the `purrr` syntax, and full `purrr` seems overkill here.

krlmlr commented on R/deparse-data-frame.R:
```diff
@@ -0,0 +1,137 @@
+#' @export
+deparse.data.frame <-
+  function(x, as_tibble = FALSE, as_tribble = FALSE, generate_mutate = TRUE, ...) {
```
I'm fine with `as_tribble = TRUE` as default.

NikNakk replied:
The problem with `as_tribble` is that after the discussion you, I and Hadley had regarding extending the `tribble` syntax, it's relatively harder to use `tribble` for data-types beyond the basic ones. I've got some code which generates a piped call to `mutate` for `factor`s, `Date`s, etc., but it's not ideal.

krlmlr replied:
That's a good point, we can decide later. The `as_tribble` format also needs better vertical alignment.

krlmlr commented on R/deparse-data-frame.R:
```diff
@@ -0,0 +1,137 @@
+  )
+
+  output_collapsed <- map_chr(
+    1:nrow(x),
```
`seq_len(nrow(x))` avoids corner cases.

NikNakk replied:
Changed.

krlmlr commented on R/deparse-data-frame.R:
```diff
@@ -0,0 +1,137 @@
+        "\n  )"
+      )
+    } else {
+      warning("Without `generate_mutate`, deparsed code may not function correctly on types such as factors")
```
If we have rlang, we can use `warn()` and `abort()` consistently.

NikNakk replied:
Changed.

krlmlr replied:
I still see `warning()` in the line above.

krlmlr commented on README.Rmd:
```diff
@@ -14,6 +14,8 @@ knitr::opts_chunk$set(
 
 # deparse
 
+[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/NikNakk/deparse?branch=master&svg=true)](https://ci.appveyor.com/project/NikNakk/deparse)
```
Does this need to point to the source repo? Also, could you please add the Travis badge?

NikNakk replied:
Yes, it should. I'll add travis too.

krlmlr commented on tests/testthat/test-deparse.R:
```diff
@@ -0,0 +1,78 @@
+test_that("deparse handles functions", {
+  check_deparse_equal(function(x) x + 1)
+  check_deparse_equal(mean)
+  expect_output(print(deparsec(function(x) x + 1)), "function \\(x\\) x \\+ 1")
```
I'd prefer `fixed = TRUE` here and elsewhere. Any reason not to use `check_deparse_equal()`?

NikNakk replied:
It breaks on some tests e.g.

    check_deparse_equal(tibble(x = 1:3, y = list(4:6, 7:9, 10:15)))
    # Error in equal_data_frame(target, current, ignore_col_order = ignore_col_order,  : 
    #  Can't join on 'y' x 'y' because of incompatible types (list / list)

krlmlr reviewed (commented):
Thanks, this is awesome! I haven't reviewed all of the new code, but from the tests it looks like a huge leap forward.

I have enabled Travis-CI and AppVeyor for this project.

In general, I'd prefer constructing language objects  instead of pasting them together, but this requires a way to represent whitespace (or at least newlines) inside a call object.

krlmlr commented:
Perhaps:

``` r
x <- as.call(list(quote(fun), structure(1, some_attrib = 42), 2))
x
#> fun(1, 2)
as.list(x)
#> [[1]]
#> fun
#> 
#> [[2]]
#> [1] 1
#> attr(,"some_attrib")
#> [1] 42
#> 
#> [[3]]
#> [1] 2
as.list(x)[[2]]
#> [1] 1
#> attr(,"some_attrib")
#> [1] 42
```

Commit:
Fixed various bits from PR #4

- `rlang` Depends changed to Imports
- travis badge added
- badges changed to source repos
- tests changed to use `fixed = TRUE`

NikNakk commented:
Thanks @krlmlr. I've made adjustments as indicated above. Clearly quite a bit of work still needed, including updated documentation, examples, etc.

krlmlr commented on DESCRIPTION:
```diff
@@ -1,7 +1,7 @@
 Title: What the Package Does (one line, title case)
 Version: 0.0-0
-Authors@R: person("Kirill", "Müller", role = c("aut", "cre"), email = "krlmlr+r@mailbox.org")
+Authors@R: person("Kirill", "M�ller", role = c("aut", "cre"), email = "krlmlr+r@mailbox.org")
```
The multibyte character was introduced here. DESCRIPTION should be in UTF-8, as stated in the `Encoding` field.

krlmlr commented:
I have merged half of this PR which integrates AppVeyor and Travis, and fixes R CMD check issues. Could you please rebase and force-push?

NikNakk commented:
I've rebased and force pushed.

Commit:
Updated authors in DESCRIPTION

Added NK, tried again with u umlaut

Commit:
Align columns in tribbles

Also added warning for tribbles with row names

krlmlr commented on DESCRIPTION:
```diff
@@ -1,22 +1,27 @@
-Encoding: UTF-8
 LazyData: true
 Date: 2016-08-16
 URLNote: https://github.com/krlmlr/
```
The repo name is missing here and in the two lines below.

krlmlr commented on R/deparse-data-frame.R:
```diff
@@ -1,20 +1,22 @@
+deparse.data.frame <-
+  function(x, as_tibble = FALSE, as_tribble = FALSE, generate_mutate = TRUE, ...) {
+
+    need_row_names <- any(row.names(x) != as.character(seq_len(nrow(x))))
```
Indention looks wrong here.

krlmlr replied:
Also, you could use the code of `tibble::has_row_names()` to detect existence of row names.

krlmlr commented on R/deparse-list.R:
```diff
@@ -1,26 +1,52 @@
+    add_class <- ""
+  }
+  add_attr_names <- setdiff(names(x_attrs), exclude_attrs)
+  add_attr_labels <- stats::setNames(add_attr_names, add_attr_names)
```
`rlang` also has `set_names()`

krlmlr reviewed (commented):
Thanks. I've noticed that the closing paren in `as_tribble = TRUE` needs to be unindented, and that in `as_tribble = FALSE` each column should be on its own line. I'd like to let this PR stay open for a while and play with it before merging, please feel free to add commits at your convenience.

krlmlr commented:
Thanks!
